### PR TITLE
Send back response when timeout

### PR DESCRIPTION
**Pull request by riyafa**
When there is a connection timeout currently the channel is closed. In this PR if the timeout is during the reception of a request or after completion of the reception of the request a response is sent back to inform the client. If it is the timeout when reading the request 408 Request Timeout is sent. If it is the timeout before writing a response a 500 internal server error is sent.

Resolve ballerina-platform/ballerina-lang#3794

### CODE REVIEW

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java [hunk 1]
```diff
@@ -245,20 +258,29 @@ private HTTPCarbonMessage setupCarbonMessage(HttpMessage httpMessage, ChannelHan
...
         sourceErrorHandler.exceptionCaught(cause);
     }
 
     @Override
-    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
         if (evt instanceof IdleStateEvent) {
             this.idleTimeout = true;
-            this.channelInactive(ctx);
+            if (sourceErrorHandler.getState() == SourceInteractiveState.RECEIVING_ENTITY_BODY) {
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java]: Shall we move this code to sourceErrorHandler? Intention is to handle all the error scenarios from a single class.

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java [hunk 2]
```diff
@@ -313,6 +313,9 @@
     public static final String REMOTE_SERVER_CLOSE_RESPONSE_CONNECTION_AFTER_REQUEST_READ
             = "Remote host closed the connection without sending inbound response";
 
+    public static final String CHANNEL_CLOSED = "Connection closed when writing response";
+
+
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java]: Shall we remove this extra new line?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceErrorHandler.java [hunk 3]
```diff
@@ -99,12 +114,19 @@ void handleIdleErrorScenario(HTTPCarbonMessage inboundRequestMsg, HttpResponseFu
                     break;
                 case RECEIVING_ENTITY_BODY:
                     handleIncompleteInboundRequest(IDLE_TIMEOUT_TRIGGERED_WHILE_READING_INBOUND_REQUEST);
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceErrorHandler.java]: Notifying server connector should happen at last after the connection close. Other wise sever response will return before Timeout response

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java [hunk 4]
```diff
@@ -313,6 +313,8 @@
     public static final String REMOTE_SERVER_CLOSE_RESPONSE_CONNECTION_AFTER_REQUEST_READ
             = "Remote host closed the connection without sending inbound response";
 
+    public static final String CHANNEL_CLOSED = "Connection closed when writing response";
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/common/Constants.java]: Shall we rephrase this as "Connection closed while writing outbound response"

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceErrorHandler.java [hunk 5]
```diff
@@ -96,13 +110,19 @@ void handleIdleErrorScenario(HTTPCarbonMessage inboundRequestMsg, HttpResponseFu
...
-                    serverConnectorFuture.notifyErrorListener(
-                            new ServerConnectorException(IDLE_TIMEOUT_TRIGGERED_BEFORE_INITIATING_OUTBOUND_RESPONSE));
+                    if (evt.state() != IdleState.READER_IDLE) {
+                        //500 Internal Server error is sent if the idle timeout is reached
+                        return sendRequestTimeoutResponse(ctx, HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                                                          copiedBuffer("Server time out", CharsetUtil.UTF_8));
+                    }
+                    serverConnectorFuture.notifyErrorListener(new ServerConnectorException(
+                            IDLE_TIMEOUT_TRIGGERED_BEFORE_INITIATING_OUTBOUND_RESPONSE));
+
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceErrorHandler.java]: Shall we remove this new line?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java [hunk 6]
```diff
@@ -246,20 +245,37 @@ private HTTPCarbonMessage setupCarbonMessage(HttpMessage httpMessage, ChannelHan
...
+                outboundRespFuture.addListener((ChannelFutureListener) channelFuture -> {
+                    sourceErrorHandler.setState(ENTITY_BODY_SENT);
+                    Throwable cause = channelFuture.cause();
+                    if (cause != null) {
+                        log.warn("Failed to send: " + cause.getMessage());
+                    }
+                    this.channelInactive(ctx);
+                    if (inCompleteRequest) {
+                        sourceErrorHandler.handleIncompleteInboundRequest(
+                                IDLE_TIMEOUT_TRIGGERED_WHILE_READING_INBOUND_REQUEST);
```
Comment (chamil321) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java]: Shall we remove this new line?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java [hunk 7]
```diff
@@ -246,20 +244,36 @@ private HTTPCarbonMessage setupCarbonMessage(HttpMessage httpMessage, ChannelHan
...
+            if (outboundRespFuture == null) {
+                this.channelInactive(ctx);
+            } else {
+                outboundRespFuture.addListener((ChannelFutureListener) channelFuture -> {
+                    sourceErrorHandler.setState(ENTITY_BODY_SENT);
+                    Throwable cause = channelFuture.cause();
+                    if (cause != null) {
+                        log.warn("Failed to send: " + cause.getMessage());
+                    }
+                    this.channelInactive(ctx);
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java]: Can't we move the below to here ?

boolean inCompleteRequest = sourceErrorHandler.getState() == RECEIVING_ENTITY_BODY;

Comment (riyafa) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/listener/SourceHandler.java]: Nope. I am changing the states when sending the message and after message is sent.
